packages/docsify-server-renderer/index.js

```diff
@@ -1,18 +1,34 @@
 import { readFileSync } from 'fs';
 import { resolve, basename } from 'path';
 import resolvePathname from 'resolve-pathname';
+import fetch from 'node-fetch';
+import debug from 'debug';
+import DOMPurify from 'dompurify';
 import { AbstractHistory } from '../../src/core/router/history/abstract';
 import { Compiler } from '../../src/core/render/compiler';
 import { isAbsolutePath } from '../../src/core/router/util';
 import * as tpl from '../../src/core/render/tpl';
 import { prerenderEmbed } from '../../src/core/render/embed';
-import fetch from 'node-fetch';
-import debug from 'debug';
 
 function cwd(...args) {
   return resolve(process.cwd(), ...args);
 }
 
+// Borrowed from https://j11y.io/snippets/getting-a-fully-qualified-url.
+function qualifyURL(url) {
+  const img = document.createElement('img');
+  img.src = url; // set string url
+  url = img.src; // get qualified url
+  img.src = ''; // prevent the server request
+  return url;
+}
+
+function isExternal(url) {
+  url = qualifyURL(url);
+  url = new URL(url);
+  return url.origin !== location.origin;
```

Comment:
same for `location`, it will be undefined in server !

Reply:
> All checks have passed

I suppose we need testing for server side. Or does it have `document` in the test env perhaps?

Reply:
ditto

Reply:
we dont have test in server side. We need to manually test it in our machine

Reply:
We can use this in Node.js: https://nodejs.org/api/url.html#url_url_resolve_from_to

Reply:
But what does `isExternal` mean on the server side, if it does not know what the origin is?

Reply:
the server will get the absolute link (URL) of the request path. so we can only allow if the origin is the same as the server. 

In my original commit, it what setup like that to validate the link

https://github.com/docsifyjs/docsify/pull/1128/commits/1e3dd13095c28d5815d908d618b5aae708f34bda#diff-dcea65610b6f84686083cbc6ea82dcb2R17-R42

Reply:
Here's another idea: What if we use `jsdom` on the server side, which not only includes `img`, `document`, `location`, but also has `DOMParser` which we can replace `DOMPurify` with? Then the purification code (our use of DOMParser) will be the same on the client and server, and the client will be without the DOMPurify dependency.

Reply:
we only need to purify/sanitize the remote URLs

Reply:
JSDOM is quite a big package.

https://bundlephobia.com/result?p=jsdom@16.2.2.

Reply:
Sorry, GitHub updates are slow, I just refreshed and see all your other comments.

Looking at dompurify devDependencies, it already uses `jsdom` for the Node.js side. https://github.com/cure53/DOMPurify/blob/master/package.json#L72

> we dont have test in server side. We need to manually test it in our machine

I haven't used any of the server side stuff. What should I do to test it?

> the server will get the absolute link (URL) of the request path. so we can only allow if the origin is the same as the server.

Ah ok. If we do know the origin, I think `url.resolve` will be useful for it then.

Reply:
Confirmed: if you look at this file, https://unpkg.com/dompurify@2.0.11/dist/purify.cjs.js, that's the one that gets imported in Node.js, and it includes `DOMParser` from `jsdom`.

Reply:
So if we load that on the server, then `window`, `location`, `document`, etc, will all be there.

Reply:
I think as an exercise I'd like to write a test for this, to make sure.

Reply:
DOMParser is a purifier ? or its the parser ? I think we need to manually purify it right ?

> Then the purification code (our use of DOMParser) will be the same on the client and server, and the client will be without the DOMPurify dependency.

also, can you explain this, I didn't get it. SSR is optional so if user is not using ssr then we still need to purify the links in client

Reply:
a clarification on ⬆️ would be great and yea,  it seems github is not working properly to load the comments , I too have to manually refresh it. 


yes, test for SSR would be awesome. 

there is an issue will router and SSR because of `location` undefined so I think we can use `JSDOM` in the server but still we need `DOMpurifier` for client I think.

Reply:
> DOMParser is a purifier ? or its the parser ? I think we need to manually purify it right ?

After some more research, turns out DOMPurify is quite simple, and is written on top of DOMParser. DOMPurify basically has a list of all the allowed attributes and tags, and removes any that aren't allowed.

For example, [attrs.js](https://github.com/cure53/DOMPurify/blob/master/src/attrs.js)

And they make a whitelist here: https://github.com/cure53/DOMPurify/blob/master/src/purify.js#L167-L183

Seems they could make the lib smaller if they instead made a blacklist. I would assume there are many more allowed tags/attributes than disallowed tags/attributes. However a blacklist is more insecure, for example if new tags or attributes are added then they might slip through, but with a white list all new tags/attrs are ignored.

> SSR is optional so if user is not using ssr then we still need to purify the links in client

I just meant that if we used DOMParser manually on both sides, then that could work too. But turns out DOMPurify is basically doing exactly what I was thinking to do.

Reply:
So we are going to keep both client and server same right ? with using DOMpurifier ?

Reply:
Yeah, sounds good, especially if we'll also use jsdom on the serverside for to get plugins working in SSR, and static rendering. :+1: Ok, I'm off for tonight!